File: py-algorithm/number/valid_square.py
import unittest

from typing import List

"""
    https://leetcode.com/problems/valid-square/

    Cho tọa độ của 4 điểm, kiểm xem chúng có tạo được hình vuông hay không?
"""


class Solution(unittest.TestCase):

    # Tính bình phương khoảng cách của 2 điểm
    def distance_square(self, A: List[int], B: List[int]):
        return (A[0] - B[0]) ** 2 + (A[1] - B[1]) ** 2

    def validSquare(self, p1: List[int], p2: List[int], p3: List[int], p4: List[int]) -> bool:
        unique = set()
        unique.add(self.distance_square(p1, p2))
        unique.add(self.distance_square(p1, p3))
        unique.add(self.distance_square(p1, p4))
        unique.add(self.distance_square(p2, p3))
        unique.add(self.distance_square(p2, p4))
        unique.add(self.distance_square(p3, p4))
        # Nếu 4 điểm tạo thành hình vuông thì unique chỉ chứa 2 số là bình phương
        # độ dài cạnh và canh huyền.
        return 0 not in unique and len(unique) == 2

    def test(self):
        self.assertEqual(True, self.validSquare([0, 0], [1, 1], [1, 0], [0, 1]))
        self.assertEqual(False, self.validSquare([0, 0], [1, 1], [1, 0], [0, 12]))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/java/number/ValidSquare.java
package number;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/*
    https://leetcode.com/problems/valid-square/

    Cho tọa độ của 4 điểm, kiểm xem chúng có tạo được hình vuông hay không?
 */
public class ValidSquare {

    // Tính bình phương khoảng cách của 2 điểm
    public int distanceSquare(int[] a, int[] b) {
        return (a[0] - b[0]) * (a[0] - b[0]) + (a[1] - b[1]) * (a[1] - b[1]);
    }

    public boolean validSquare(int[] p1, int[] p2, int[] p3, int[] p4) {
        Set<Integer> set = new HashSet<>();
        set.add(distanceSquare(p1, p2));
        set.add(distanceSquare(p1, p3));
        set.add(distanceSquare(p1, p4));
        set.add(distanceSquare(p2, p3));
        set.add(distanceSquare(p2, p4));
        set.add(distanceSquare(p3, p4));
        /**
         * - Không chứa số 0 vì khi đó có 2 điểm trùng nhau.
         * - Set chỉ chứa bình phương của 4 cạnh và 2 đường chéo nên size = 2.
         */
        return !set.contains(0) && set.size() == 2;
    }

    @Test
    public void test() {
        assertEquals(true, validSquare(new int[]{0, 0}, new int[]{1, 1}, new int[]{1, 0}, new int[]{0, 1}));
    }

}
